core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java

```diff
@@ -546,10 +565,52 @@ private RelationPlan createInsertPlan(Analysis analysis, Insert insertStatement)
         TableHandle tableHandle = insert.getTarget();
         Query query = insertStatement.getQuery();
         Optional<TableLayout> newTableLayout = insert.getNewTableLayout();
-        return getInsertPlan(analysis, insert.getTable(), query, tableHandle, insert.getColumns(), newTableLayout, Optional.empty());
+        return getInsertPlan(analysis, Optional.of(insert.getTable()), query, tableHandle, insert.getColumns(), newTableLayout, Optional.empty());
+    }
+
+    private RelationPlan createMaterializedViewCreationPlan(Analysis analysis)
+    {
+        checkState(analysis.getCreateMaterializedView().isPresent(), "CreateMaterializedViewAnalysis handle is missing");
+        Analysis.CreateMaterializedViewAnalysis create = analysis.getCreateMaterializedView().get();
+        createMaterializedView(create);
+
+        Optional<MaterializedViewDefinition> definition = metadata.getMaterializedView(session, create.getViewName());
+        if (definition.isEmpty()) {
+            throw new TrinoException(GENERIC_INTERNAL_ERROR, "Materialized view metadata is not available after it is created");
+        }
+        CatalogSchemaTableName storageTable = definition.get().getStorageTable()
+                .orElseThrow(() -> new TrinoException(GENERIC_INTERNAL_ERROR, "Materialized view storage table is empty"));
+        TableHandle tableHandle = metadata.getTableHandle(session,
+                        new QualifiedObjectName(storageTable.getCatalogName(),
+                                storageTable.getSchemaTableName().getSchemaName(),
+                                storageTable.getSchemaTableName().getTableName()))
+                .orElseThrow(() -> new TrinoException(GENERIC_INTERNAL_ERROR, "Materialized view storage table is empty"));
+        Map<String, ColumnHandle> columnHandles = metadata.getColumnHandles(session, tableHandle);
+        List<ColumnHandle> columns = metadata.getTableMetadata(session, tableHandle).getColumns().stream()
+                .filter(column -> !column.isHidden())
+                .map(columnMetadata -> columnHandles.get(columnMetadata.getName()))
+                .collect(toImmutableList());
+
+        return createRefreshMaterializedViewPlanInternal(analysis, Optional.empty(), tableHandle, create.getQuery(), columns);
+    }
+
+    private void createMaterializedView(Analysis.CreateMaterializedViewAnalysis create)
+    {
+        // TODO: This is not an ideal place to create a materialized view metadata and the corresponding storage table.
+        //       However, since currently storage table creation is completely hidden inside a connector interface, relevant information on the storage table cannot be retrieved without really creating it.
+        //       With more communications between planner and connector by introducing new interfaces, the storage table information such as table name, properties, and table layout can be retrieved without creating it.
```

Comment:
Do you have some idea on SPI for that? Maybe there should be something similar as `io.trino.spi.connector.ConnectorMetadata#getNewTableLayout`, but for MVs.

Reply:
I am working on extending SPI to remove this code here. We can further discuss it in a separate PR because this PR is already quite big.